### ISSUE TITLE
Renewal Issue with -AllAccounts

### ISSUE THREAD

**Issue by thecarnie** [rmbolger/Posh-ACME#344]
I am not sure if I'm doing something wrong - but this process was working in the past for our configuration.

Breakdown:
Running Windows Server 2019 + IIS + MS-DNS.
Running our scripts to request initial certificates works properly.
Running our script to call the renewals task every 12 hours for expired certs fails with no "renewal orders" found.

I currently have over a dozen expired certificates and none of them are showing "ready to renew" - but here is one of the order.json files from the suspect domains.

I am needing to know what further information you might need - and if there is something I'm missing on our renewals script.

(attaching both)

**SAMPLE OF AN order.json FROM PROBLEM RENEWAL** (some data masked)
```json
{
    "status":  "pending",
    "expires":  "2021-05-15T08:50:11Z",
    "identifiers":  [
                        {
                            "value":  "domainname.tld",
                            "type":  "dns"
                        },
                        {
                            "value":  "*.domainname.tld",
                            "type":  "dns"
                        }
                    ],
    "authorizations":  [
                           "https://acme-v02.api.letsencrypt.org/acme/authz-v3/######",
                           "https://acme-v02.api.letsencrypt.org/acme/authz-v3/######"
                       ],
    "finalize":  "https://acme-v02.api.letsencrypt.org/acme/finalize/######/######",
    "OCSPMustStaple":  false,
    "DnsSleep":  120,
    "PreferredChain":  "",
    "AlwaysNewKey":  false,
    "DnsAlias":  [
                     "",
                     ""
                 ],
    "MainDomain":  "domainname.tld",
    "FriendlyName":  "domainname.tld",
    "CertExpires":  null,
    "KeyLength":  "2048",
    "RenewAfter":  null,
    "Install":  true,
    "UseSerialValidation":  false,
    "Plugin":  [
                   "Windows",
                   "Windows"
               ],
    "SANs":  [
                 "*.domainname.tld"
             ],
    "ValidationTimeout":  60,
    "certificate":  null,
    "location":  "https://acme-v02.api.letsencrypt.org/acme/order/#######/#######",
    "PfxPassB64U":  "xxxxxxxxxxx"
}
```

And the **RENEWAL SCRIPT**
```powershell
Set-Location C:\automation

Set-PAServer LE_PROD

Submit-Renewal -AllAccounts -Verbose | ForEach-Object {
    $cert = $_
    # deploy for everything else
    $CN = "$cert.MainDomain"
    $TLD = $CN.Replace("*.","")

    $cert | Set-IISCertificate -SiteName "xxxxxxxxx" -HostHeader "webmail.$TLD" -RequireSNI -Verbose -RemoveOldCert
    $cert | Set-IISCertificate -SiteName "xxxxxxxxx" -HostHeader "autodiscover.$TLD" -RequireSNI -Verbose -RemoveOldCert
    Write-Host "$TLD Attempted Renew Bindings!"
    

}
```

**I thank you greatly for your time on this issue** - as I fear it is a configuration issue on my end, but I can't seem to pin it down.

**Comment by thecarnie** [comment]
An addendum - submitting manual renewals one domain at a time works as expected using the following script:

```powershell
Param(
    [Parameter(Mandatory=$true)]
    [ValidateNotNullOrEmpty()]
    [string] $Domain
    
)

Set-Location C:\automation

Set-PAServer LE_PROD

Set-PAOrder $Domain

if ($cert = Submit-Renewal) {
    $cert | Set-IISCertificate -SiteName "xxxxxxx" -HostHeader "webmail.$Domain" -RequireSNI -Verbose -RemoveOldCert
    $cert | Set-IISCertificate -SiteName "xxxxxxx" -HostHeader "autodiscover.$Domain" -RequireSNI -Verbose -RemoveOldCert
    Write-Host "$TLD Attempted Renew Bindings!"
}

```

**Comment by rmbolger** [comment]
Hey @thecarnie. Happy to try and help figure this out with you.

The `pending` status of the sample order makes it seem like an error might be interrupting the renewal process somewhere. When you run the renewal script for all accounts are there any errors displayed in the output? Is there a (sanitized) copy of the output you could post?

**Comment by thecarnie** [comment]
I got a logfile from the last "renewals" run -- (and a little more background on our configuration):

Long story short - our SSLs are DNS authorized from a Master Windows DNS server we control - and credentials to set the DNS TXT records are stored in each cert's data folder (pluginargs.json) when we ordered the initial certificates.  When renewals tries to run - it behaves like it can't find either the pluginargs.json file or it can't find the credentials it needs to renew the domain.  

**NOTE** our system relies on the POSHACME_HOME environment variable (set at the SYSTEM level, not USER) - which seems to work properly or our "ordering" of initial certs would not store properly where we have them directed - so I don't think this is the issue - or POSH wouldn't be able to even find the orders it's trying to renew.  It almost seems like it's an issue reading the pluginargs.json file (or finding it).

**from the logfile*** (there were about 10 domains to renew in that run - all failed same errors)
```powershell
Renewing certificate for order domainname.tld
ConvertTo-SecureString : The system cannot find the path specified.

At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.0.0\Public\Get-PAPluginArgs.ps1:61 char:70
+ ... prop.Value.user,($prop.Value.pass | ConvertTo-SecureString @encParam)
+                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertTo-SecureString], CryptographicException
    + FullyQualifiedErrorId : ImportSecureString_InvalidArgument_CryptographicError,Microsoft.PowerShell.Commands.Conv 
   ertToSecureStringCommand
 
Get-PAPluginArgs : Cannot process argument because the value of argument "password" is null. Change the value of 
argument "password" to a non-null value.
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.0.0\Public\Submit-Renewal.ps1:89 char:42
+ ...           $certParams.PluginArgs = Get-PAPluginArgs $order.MainDomain
+                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-PAPluginArgs], PSArgumentNullException
    + FullyQualifiedErrorId : PSArgumentNullException,Get-PAPluginArgs
 
Using ACME Server https://acme-v02.api.letsencrypt.org/directory
Using account 107632508
Creating a new order for domainname.tld, *.domainname.tld
ConvertTo-SecureString : The system cannot find the path specified.

At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.0.0\Public\Get-PAPluginArgs.ps1:61 char:70
+ ... prop.Value.user,($prop.Value.pass | ConvertTo-SecureString @encParam)
+                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [ConvertTo-SecureString], CryptographicException
    + FullyQualifiedErrorId : ImportSecureString_InvalidArgument_CryptographicError,Microsoft.PowerShell.Commands.Conv 
   ertToSecureStringCommand
 
Get-PAPluginArgs : Cannot process argument because the value of argument "password" is null. Change the value of 
argument "password" to a non-null value.
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.0.0\Public\Submit-ChallengeValidation.ps1:90 char:23
+         $PluginArgs = Get-PAPluginArgs $Order.MainDomain
+                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-PAPluginArgs], PSArgumentNullException
    + FullyQualifiedErrorId : PSArgumentNullException,Get-PAPluginArgs
 
Publishing challenge for Domain domainname.tld with Token ybZdx_8DejHmu_vljM8sutaNUdYWA5CXz6m2E8Dw8mk 
using Plugin Windows and DnsAlias ''.
Submit-ChallengeValidation : Cannot bind argument to parameter 'WinServer' because it is an empty string.
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.0.0\Public\New-PACertificate.ps1:216 char:9
+         Submit-ChallengeValidation
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Submit-ChallengeValidation], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorEmptyStringNotAllowed,Submit-ChallengeValidation
```

What's interesting is if I run them one by one - there is no issue.  Additionally, when I run an order manually (which is common because Plesk doesn't support our configuration for SSL) - there is no issue creating the initial orders.

**order.json** from the above renewal entry
```json
{
    "status":  "pending",
    "expires":  "2021-05-15T21:04:00Z",
    "identifiers":  [
                        {
                            "value":  "domainname.tld",
                            "type":  "dns"
                        },
                        {
                            "value":  "*.domainname.tld",
                            "type":  "dns"
                        }
                    ],
    "authorizations":  [
                           "https://acme-v02.api.letsencrypt.org/acme/authz-v3/12960052782",
                           "https://acme-v02.api.letsencrypt.org/acme/authz-v3/12960052781"
                       ],
    "finalize":  "https://acme-v02.api.letsencrypt.org/acme/finalize/107632508/9570586750",
    "OCSPMustStaple":  false,
    "DnsSleep":  40,
    "PreferredChain":  "",
    "AlwaysNewKey":  false,
    "DnsAlias":  [
                     "",
                     ""
                 ],
    "MainDomain":  "domainname.tld",
    "FriendlyName":  "domainname.tld",
    "CertExpires":  null,
    "KeyLength":  "2048",
    "RenewAfter":  null,
    "Install":  true,
    "UseSerialValidation":  false,
    "Plugin":  [
                   "Windows",
                   "Windows"
               ],
    "SANs":  [
                 "*.domainname.tld"
             ],
    "ValidationTimeout":  60,
    "certificate":  null,
    "location":  "https://acme-v02.api.letsencrypt.org/acme/order/107632508/9570586750",
    "PfxPassB64U":  "xxxxxxxx"
}
```

**pluginargs.json** from same referenced renewal
```json
{
    "WinUseSSL":  true,
    "WinCred":  {
                    "origType":  "pscredential",
                    "user":  "DOMAIN\\authuser",
                    "pass":  "masked----scrubbed----masked----scrubbed---masked"
                },
    "WinServer":  "masterdns.domainname.tld"
}
```


Thank you very much for your time in this matter.

**Comment by thecarnie** [comment]
UNRELATED:  In the midst of running today's renewals one by one - after the 13th cert, now Let's Encrypt is returning this:

```text
Update-PAServer : The remote server returned an error: (503) Server Unavailable.
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.4.0\Public\Set-PAServer.ps1:61 char:9
+         Update-PAServer @updateParams
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WebException
    + FullyQualifiedErrorId : System.Net.WebException,Update-PAServer
 
ACME response body was not JSON parseable
The remote server returned an error: (405) Method Not Allowed.
At C:\Program Files\WindowsPowerShell\Modules\Posh-ACME\4.4.0\Private\Invoke-ACME.ps1:153 char:17
+                 throw $ex
+                 ~~~~~~~~~
    + CategoryInfo          : OperationStopped: (:) [], WebException
    + FullyQualifiedErrorId : The remote server returned an error: (405) Method Not Allowed.

```
I believe this is a rate-limiting issue, but the other day when I opened this ticket, I ran some 30 renewals without issue...

(Also -- I'm running Posh-ACME v4.4.0 - just as an aside)

**Comment by rmbolger** [comment]
Looks like Let's Encrypt is having a planned maintenance window at the moment which would be why you can't connect. Rate limit errors would theoretically show a more descriptive, ACME specific error.

https://letsencrypt.status.io/

**Comment by thecarnie** [comment]
That was my theory about rate-limiting and service unavailable - as we have never hit that one before, and my other log files don't show any of those under normal operation.

**Comment by thecarnie** [comment]
I am going to investigate this -- currently, the entire process is run by a single user reserved strictly for the cert order and renewal process -- this has been from the beginning due to weird issues I ran into (basically what you described above) when originally working on my scripts to handle the process.  Once I saw issues in the credentials, we started over with a single "certuser" account with necessary privileges for working with IIS and MS-DNS remotely (essentially a somewhat limited admin account). 

That said - the concept of using a "universal" AES key (local) vs. OS controlled seems much more stable in the long run regardless - because disaster can strike, and if user accounts had to be re-created - all those keys are toast and need re-initialized.

Thank you for the additional information! I look forward to seeing if there is an issue with the system being unable to find the `pluginargs.json` file as "File Not Found Exception" seems to point in that direction.
